### DISCUSSION TITLE
ALOS-2 batch processing with common reference image

### DISCUSSION THREAD

**Discussion by leomarconato** [gmtsar/gmtsar#872] [Q&A] [answered]
Hi,

First, thanks to the developers for the work done on this processing chain, which is really easy to get familiar with.

I have a bunch of ALOS-2 ScanSAR acquisitions (~40) for the same frame, and I successfully computed interferograms from pairs using _p2p_ALOS2_SCAN_Frame.csh_.

Now, I would like to derive an SBAS time-series from this dataset, and the first step would be to generate interferograms in the geometry of a single reference image.
To do so, I tried to build a workflow based on GMTSAR's scripts and following the logic of _p2p_ALOS2_SCAN_Frame.csh_ :
       1. Pre-process all acquisitions with _pre_proc_batch_ALOS2_SCAN.csh_ + resample all subswaths at the same PRF (_samp_slc.csh_).
       2. For each subswath, align all the SLCs with _p2p_processing.csh_ (step 2 only), with the same reference image for all.
       3. Define pairs of images and compute interferogram (_intf.csh_+_filter.csh_) for each subswath.
       4. Merge all subswaths for each interferogram, using the same strategy as in _p2p_ALOS2_SCAN_Frame.csh_, i.e. using _merge_unwrap_geocode_tops.csh_ twice, basically.
 
However, I end up with merged interferograms featuring very heterogeneous coherence between subswaths, and sometimes large ramps in one subswath and not the other (there is also a lot of ionospheric noise, but this will be a next step) => attached file. This makes me think of a problem of coregistration between some sub-swath SLCs. I tried to change the reference image, but the results are still bad.
Thus, I wondered if anyone already processed such data, and if my processing strategy was wrong maybe, or if there is another problem (ionospheric noise preventing alignment between two SLCs very spaced in time... ?).

Thank you in advance for your help!

![all_phasefilt_small](https://github.com/gmtsar/gmtsar/assets/73933829/f8b3e8b1-a951-4913-bf06-354be644c938)

**Comment by Xiaohua-Eric-Xu** [comment]
@ericlindsey might be able to provide a better answer. 

For my personal opinion, coregistration on subswaths may get you different parameters. But normally they would not provide a very different ramp unless the coherence is terrible. 

If p2p_ALOS2_SCAN_Frame.csh is able to provide you a good interferogram, then the batch processing gets you an unfamiliar solution, there must be something incorrect with the batch processing. However, the strategy you provided looks good to me. 

Since seems you are quite experienced with GMTSAR, here are some other choices. There is a recently implemented skip_master option built for batch processing into p2p_processing.csh. The function is now built into a script called batch_processing.csh but not for ALOS2 ScanSAR. You could modify the script by adding the samp_slc.csh step. Then use batch_processing.csh to process every subswaths, merge, unwarp... I can't guarantee that this'll work perfectly though.

Another scenario is that the coregistration may have get to a scattered peak when coherence is not very good. See the scatter for a single reflector in ScanSAR data below. We have some coregistration step to avoid this but one has to specify correctly by using mode ALOS2_SCAN for p2p_processing.csh 
![image](https://github.com/gmtsar/gmtsar/assets/16844111/dfda0b88-6748-4e44-b023-d03431412d1c)

**Reply by ericlindsey** [reply, marked as answer]
Do some of the swaths contain more ocean areas than others? This definitely looks like a coregistration issue to me, possibly related to @Xiaohua-Eric-Xu 's figure showing the multiple peaks associated with point reflectors. This can be a tricky thing to solve - you may need to adjust some of the parameters in the alignment step. The lines to think about are here:

          xcorr $master.PRM $aligned.PRM -xsearch 32 -ysearch 256 -nx 32 -ny 128
          awk '{print $4}' < freq_xcorr.dat > tmp.dat
          set amedian = `sort -n tmp.dat | awk ' { a[i++]=$1; } END { print a[int(i/2)]; }'`
          set amax = `echo $amedian | awk '{print $1+3}'`
          set amin = `echo $amedian | awk '{print $1-3}'`
          awk '{if($4 > '$amin' && $4 < '$amax') print $0}' < freq_xcorr.dat > freq_alos2.dat
          fitoffset.csh 2 3 freq_alos2.dat 10 >> $aligned.PRM

One easy thing to try here would be to increase the nx and ny values in the xcorr step, then increase the value of 'SNR' in the fitoffset step (currently set to 10). This will improve the number of estimates of the alignment, and keep only the better values, at the cost of longer run time. 

It might also be useful to grab the 'freq_xcorr.dat' file for each of the subswaths, and plot the values graphically so that you can see if there is a large area that is causing the coregistration error. For example, if the southern half of subswaths 1 and 2 are consistenly mis-registered, it might be useful to throw away all those values before doing fitoffset.csh on the remaining ones. You might have to write a bit of custom code here if the simpler option above doesn't work easily.

Good luck, hopefully this works! Would love to hear what you find out.

**Comment by leomarconato** [comment]
Thank you for your answers!

After more in depth analysis, I found that I was indeed facing a coregistration issue.

First, I found that ALOS-2 full aperture products were producing "secondary lines" of azimuth offsets. I don't exactly understand the reason for this, but I think it is due to the acquisition mode. This is discussed in this paper :
https://ieeexplore.ieee.org/document/7445839
... and this GitHub issue :
https://github.com/isce-framework/isce2/issues/565

Second, looking more in detail at the offsets, I noticed other lines in azimuth offsets, with different spacings for each subswath.
So, I wondered if it could be due to some kind of aliasing during the azimuth resampling with _samp_slc.csh_, and indeed, they are not visible on offsets computed on the initial SLCs before resampling. You can see them on the figure below (x-axis is azimuth).

The fact is that for long baseline pairs, the offsets get quite noisy, and for some subswaths, the fit of the offsets cannot distinguish the real trend of azimuth offsets from all these lines.
Also, the culling by median +/- 3 px can introduce a bias if is encompasses parts of lines due to resampling (e.g. for F5 on figure below).
I think that this causing inaccurate coregistration, leading to ramps in azimuth or total loss of coherence.

So, the strategy I am now  trying to implement is the following:
     - compute the offsets on initial SLCs and fit them (using a sliding median in azimuth to better cull the central line)
     - resample SLCs at the same PRF
     - convert the offset fit parameters into the resampled geometry (conversion using the ratio PRF_new/PRF_init)
     - coregister the resampled SLCs with these new parameters

I hope this may help!

<img width="995" alt="Capture d’écran 2024-02-15 à 10 45 11" src="https://github.com/gmtsar/gmtsar/assets/73933829/32ad26c0-0d29-457b-8c30-554d754ed08a">

**Reply by Xiaohua-Eric-Xu** [reply]
This is a better strategy I think. So if I understand correctly, you are coregistering and then upsample the SLCs to a larger PRF. If there are signal with higher frequency, it should show up on the spectrum (i.e., if you do a fft for a column of data, stack a few and check on the spectrogram). Then resampling will be dangerous to do. 

I am marking this as enhancement and will try changing the code afterwards. 

However, the question remains why pairwise (compared to stack) processing does not affected. Any idea?

**Reply by leomarconato** [reply]
Actually, the strategy I used is :
    - resample the SLC, but keeping the initial raw SLC
    - compute and fit the coregistration offsets on this raw SLC (relative to the supermaster raw SLC)
    - convert the fit coefficients into the geometry of the resampled SLC (using the ratio of PRFs)
    - then, coregister the resampled SLC using these converted offsets
I hope it is clearer. 

I tried _p2p_ALOS2_SCAN_Frame.csh_ on a long temporal baseline pair (~1yr), and it showed the same problems that with the batch processing. I guess that when the aligned and the supermaster images are very different, the offsets are noisier and thus the fit is more disturbed. Also, subswaths with good coherence appear to have less issues at this step that the poorly coherent ones.
Maybe using a "tree" approach with secondary master images could be an alternative strategy to improve coregistration.
